File: todo-api-js/src/adapters/controllers/todo/list.ts
import { type UseCase } from "@domain/use-case";
import { type ListTodosInput, type ListTodosOutput } from "@application/dtos/todo/list";
import { type ListTodosRequest, type ListTodosPresenter } from "@adapters/dtos/todo/list";

export class ListTodosController<TPresenter extends ListTodosPresenter<any>> {
    constructor(
        private interactor: UseCase<ListTodosInput, ListTodosOutput>,
        private presenter: TPresenter
    ) {}

    async run(req: ListTodosRequest): Promise<ReturnType<TPresenter["present"]>> {
        const input = req.parse();
        if (input.isErr()) {
            return this.presenter.present(input);
        }

        const result = await this.interactor.exec(input.value);
        return this.presenter.present(result);
    }
}

File: todo-api-rs/src/adapters/controllers/todo/list.rs
use crate::adapters::dtos::todo::list::{ListPresenter, ListRequest, ListResponseError};
use crate::application::dtos::todo::list::{ListTodosError, ListTodosInput, ListTodosOutput};
use crate::domain::use_case::UseCase;

#[derive(Debug)]
pub struct ListTodosController<T, P> {
    interactor: T,
    presenter: P,
}

impl<T, P> ListTodosController<T, P>
where
    T: UseCase<ListTodosInput, ListTodosOutput>,
    P: ListPresenter,
{
    pub const fn new(interactor: T, presenter: P) -> Self {
        Self {
            interactor,
            presenter,
        }
    }

    pub async fn run(self, req: ListRequest) -> <P as ListPresenter>::View {
        let input = match req.parse().map_err(ListResponseError::Input) {
            Ok(input) => input,
            Err(err) => return self.presenter.present(Err(err)),
        };

        let result = self.interactor.exec(input).await.map_err(|err| match err {
            ListTodosError::Internal(src) => ListResponseError::Internal(src),
        });

        self.presenter.present(result)
    }
}
